src/plugins/discourse/mirror.js

```diff
@@ -0,0 +1,242 @@
+   */
+  constructor(db: Database, fetcher: DiscourseInterface, serverUrl: string) {
+    if (db == null) throw new Error("db: " + String(db));
+    this._db = db;
+    this._fetcher = fetcher;
+    if (db.inTransaction) {
+      throw new Error("already in transaction");
+    }
+    try {
+      db.prepare("BEGIN").run();
+      this._initialize(serverUrl);
+      if (db.inTransaction) {
+        db.prepare("COMMIT").run();
+      }
+    } finally {
+      if (db.inTransaction) {
+        db.prepare("ROLLBACK").run();
+      }
+    }
+  }
+
+  _initialize(serverUrl: string) {
+    const db = this._db;
+    // We store the config in a singleton table `meta`, whose unique row
+    // has primary key `0`. Only the first ever insert will succeed; we
+    // are locked into the first config.
+    db.prepare(
+      dedent`\
+        CREATE TABLE IF NOT EXISTS meta (
+            zero INTEGER PRIMARY KEY,
+            config TEXT NOT NULL
+        )
+      `
+    ).run();
+
+    const config = stringify({
+      version: VERSION,
+      serverUrl: serverUrl,
+    });
+
+    const existingConfig: string | void = db
+      .prepare("SELECT config FROM meta")
+      .pluck()
+      .get();
+    if (existingConfig === config) {
+      // Already set up; nothing to do.
+      return;
+    } else if (existingConfig !== undefined) {
+      throw new Error(
+        "Database already populated with incompatible server or version"
+      );
+    }
+    db.prepare("INSERT INTO meta (zero, config) VALUES (0, ?)").run(config);
+
+    const tables = [
+      dedent`\
+        CREATE TABLE topics (
+            id INTEGER PRIMARY KEY,
+            title TEXT NOT NULL,
+            timestampMs INTEGER NOT NULL,
+            authorUsername TEXT NOT NULL
+        )
+      `,
+      dedent`\
+        CREATE TABLE posts (
+            id INTEGER PRIMARY KEY,
+            timestampMs INTEGER NOT NULL,
+            authorUsername TEXT NOT NULL,
+            topicId INTEGER NOT NULL,
+            postNumber INTEGER NOT NULL,
+            replyToPostNumber INTEGER,
+            FOREIGN KEY(topicId) REFERENCES topics(id)
+        )
+      `,
+    ];
+    for (const sql of tables) {
+      db.prepare(sql).run();
+    }
+  }
+
+  topics(): $ReadOnlyArray<Topic> {
+    return this._db.prepare("SELECT * FROM topics").all();
+  }
+
+  posts(): $ReadOnlyArray<Post> {
+    return this._db.prepare("SELECT * FROM posts").all();
+  }
+
+  /**
+   * Given a TopicId and a post number, find that numbered post within the topic.
+   *
+   * Returns undefined if no such post exists.
+   */
+  findPostInTopic(topicId: TopicId, postNumber: number): ?PostId {
+    return this._db
+      .prepare(
+        dedent`\
+          SELECT id
+          FROM posts
+          WHERE topicId = :topicId AND postNumber = :postNumber
+        `
+      )
+      .pluck()
+      .get({topicId, postNumber});
+  }
+
+  async update() {
+    // TODO: Make sure this happens in a transaction?
```

Comment:
Transactions give you two primary benefits (for this discussion):

  - **Atomicity:** While the transaction is running, no other statements
    will be able to interfere with it. It will execute as if it were the
    sole client.

  - **Consistency:** The transaction will either complete entirely or
    fail entirely. In no case will an intermediate state of the
    transaction be committed to the database, even in case of program
    error or power failure. 

For instance, consider:

```javascript
/**
 * Transfer some money from one bank account to another, or fail if the
 * action would cause the debitor to be overdrawn.
 */
function giveMoney(src, dst, amount) {
  const getBalance = db
    .prepare("SELECT balance FROM accounts WHERE user = ?")
    .pluck();
  const setBalance = db.prepare(
    "UPDATE accounts SET balance = :balance WHERE user = :user"
  );
  const srcBalance = getBalance.get(src);
  const dstBalance = getBalance.get(dst);
  if (srcBalance < amount) throw new Error("Hey buddy, you got no money!");
  setBalance.run({user: src, balance: srcBalance - amount});
  setBalance.run({user: dst, balance: dstBalance + amount});
}
```

As written, this has no SQL transactions. Does it need them?

First consider the case with no concurrent writers. What happens if
power fails between the first `setBalance.run` and the second? We’ll
have lost the debitor’s money and given nothing to the creditor, with no
record of the sale! So already we need SQL transactions.

(Of course, “power failure” can also mean “you got `^C`ed”, etc.)

If you have concurrent writers, you have problems even without power
failure. If we call `giveMoney(src, dst, 10)` twice, and `src` and `dst`
each have a balance of `10`, then the following could happen:

  - Both queries execute up until reading `srcBalance`, getting value
    `10` and storing this into their respective local variables.
  - The first query runs to completion, getting a `dstBalance` of `10`,
    passing the `10 > 0` check, and storing balances of `0` and `20`.
  - The first query runs to completion, getting a `dstBalance` of `20`,
    passing the `10 > 0` check, and storing balances of `0` and `30`.

Now we’ve made money out of thin air!

What we have in this situation is a **read–modify–write.** You read from
the database, perform some local computations on that value, and write
back the result. For such an operation to be safe, you need some form of
communication, the simplest form of which is a lock or transaction.

So. Read the code in this module. Where, if anywhere, does it need
transactions? What happens if power fails after each statement? or
concurrent processes execute the code, in any possible interleavings?